enterprise/server/remote_execution/cgroup/cgroup.go

```diff
@@ -31,6 +32,134 @@ const (
 	cidPlaceholder = "{{.ContainerID}}"
 )
 
+// min/max weight values for cgroup2 weight-based control
+const (
+	minWeight = 1
+	maxWeight = 10_000
+)
+
+// Setup configures the cgroup at the given path with the given settings.
+// Any settings for cgroup controllers that aren't enabled are ignored.
+// IO limits are applied to the given block device, if specified.
+func Setup(ctx context.Context, path string, s *scpb.CgroupSettings, blockDevice *block_io.Device) error {
+	m, err := settingsMap(s, blockDevice)
+	if err != nil {
+		return err
+	}
+	enabledControllers, err := ParentEnabledControllers(path)
+	if err != nil {
+		return fmt.Errorf("read enabled controllers: %w", err)
+	}
+	for name, value := range m {
+		controller, _, _ := strings.Cut(name, ".")
+		if !enabledControllers[controller] {
+			log.CtxWarningf(ctx, "Skipping cgroup %q setting for disabled cgroup controller %q", name, controller)
+			continue
+		}
+		settingFilePath := filepath.Join(path, name)
+		if err := os.WriteFile(settingFilePath, []byte(value), 0); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ParentEnabledControllers returns the cgroup controllers that are enabled for
+// the parent cgroup of a given cgroup.
+func ParentEnabledControllers(path string) (map[string]bool, error) {
+	b, err := os.ReadFile(filepath.Join(path, "..", "cgroup.subtree_control"))
+	if err != nil {
+		return nil, err
+	}
+	fields := strings.Fields(string(b))
+	enabled := make(map[string]bool, len(fields))
+	for _, f := range fields {
+		enabled[f] = true
+	}
+	return enabled, nil
+}
+
+func settingsMap(s *scpb.CgroupSettings, blockDevice *block_io.Device) (map[string]string, error) {
+	m := map[string]string{}
+	if s == nil {
+		return m, nil
+	}
+	if s.CpuWeight != nil {
+		if s.GetCpuWeight() < minWeight {
+			return nil, fmt.Errorf("invalid cpu weight %d: must be in range [1, 10000]", s.GetCpuWeight())
+		}
+		m["cpu.weight"] = strconv.Itoa(int(s.GetCpuWeight()))
+	}
+	if s.CpuQuotaPeriodUsec != nil {
+		m["cpu.max"] = fmt.Sprintf("%d %d", s.GetCpuQuotaLimitUsec(), s.GetCpuQuotaPeriodUsec())
```

Comment:
for cgroups1, quota period must be between 1000 microseconds and 1 second (https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/6/html/resource_management_guide/sec-cpu#sec-cpu). I assume it's the same for cgroups2. I guess the limit just needs to be non-negative, or maybe bigger than 0.